### ISSUE TITLE
Error: Could not prefetch selinux_fcontext provider 'semanage': uninitialized constant Selinux

### ISSUE THREAD

**Issue by trungson** [voxpupuli/puppet-selinux#311]
## Affected Puppet, Ruby, OS and module versions/distributions

- Puppet: 6.11.1
- Module version: v3.0.0
- OS: CentOS 8

## How to reproduce (e.g Puppet code you use)

	selinux::fcontext { 'selinux-www':
		seltype => 'httpd_sys_content_t',
		pathspec => '/home/www(/.*)?',
	}

## What are you seeing

Error: Could not prefetch selinux_fcontext provider 'semanage': uninitialized constant Selinux
Error: Failed to apply catalog: uninitialized constant Selinux

## What behaviour did you expect instead

No error and context for folder changed

unconfined_u:object_r:user_home_dir_t:s0 www

## Any additional information you'd like to impart

This module works fine with selinux::port in the same code, only selinux::fcontext shows that error

**Comment by vangual** [comment]
Encountered the same error yesterday but on EL7. Turned out the the ruby binary was missing the libselinux bindings, installing the libselinux-ruby package helped.

**Comment by trungson** [comment]
I installed libselinux-ruby but still same problem
```
ruby -v
ruby 2.5.3p105 (2018-10-18 revision 65156) [x86_64-linux]

 rpm -qa|grep libselinux
libselinux-utils-2.8-6.el8.x86_64
python3-libselinux-2.8-6.el8.x86_64
libselinux-devel-2.8-6.el8.x86_64
libselinux-ruby-2.8-6.el8.x86_64
libselinux-2.8-6.el8.x86_64

rpm -qa|grep ruby
ruby-irb-2.5.3-104.module_el8.0.0+179+565e49e2.noarch
rubygems-2.7.6-104.module_el8.0.0+179+565e49e2.noarch
rubygem-openssl-2.1.2-104.module_el8.0.0+179+565e49e2.x86_64
ruby-libs-2.5.3-104.module_el8.0.0+179+565e49e2.x86_64
rubygem-did_you_mean-1.2.0-104.module_el8.0.0+179+565e49e2.noarch
rubygem-psych-3.0.2-104.module_el8.0.0+179+565e49e2.x86_64
libselinux-ruby-2.8-6.el8.x86_64
rubygem-io-console-0.4.6-104.module_el8.0.0+179+565e49e2.x86_64
rubygem-rdoc-6.0.1-104.module_el8.0.0+179+565e49e2.noarch
rubygem-json-2.1.0-104.module_el8.0.0+179+565e49e2.x86_64
rubygem-bigdecimal-1.3.4-104.module_el8.0.0+179+565e49e2.x86_64
ruby-2.5.3-104.module_el8.0.0+179+565e49e2.x86_64
```

**Comment by pillarsdotnet** [comment]
I'm seeing the same problems.  I deliberately avoid installing a system ruby, because Puppet comes with its own ruby.  Unfortunately, it is not possible to install `libselinux-ruby` without also installing a system ruby, and I'm not convinced that Puppet's ruby will work with bindings for a different ruby.

Did puppet drop support for selinux?

**Comment by oranenj** [comment]
Interesting:
```
# /opt/puppetlabs/puppet/bin/irb 
irb(main):001:0> require 'selinux'
Traceback (most recent call last):
        4: from /opt/puppetlabs/puppet/bin/irb:11:in `<main>'
        3: from (irb):1
        2: from /opt/puppetlabs/puppet/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:59:in `require'
        1: from /opt/puppetlabs/puppet/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:59:in `require'
LoadError (/opt/puppetlabs/puppet/lib/ruby/vendor_ruby/2.5.0/x86_64-linux/selinux.so: undefined symbol: rpm_execcon - /opt/puppetlabs/puppet/lib/ruby/vendor_ruby/2.5.0/x86_64-linux/selinux.so
```
Seems like a bug in Puppet if you ask me.

**Comment by oranenj** [comment]
I reported an upstream bug: https://tickets.puppetlabs.com/browse/PUP-10161

**Comment by oranenj** [comment]
For now, downgrading to 6.10.1 seems to work

**Comment by blackknight36** [comment]
Confirmed that downgrading to 6.10.1 will fix this problem. Those affected by this can use the following command to downgrade.

`dnf install puppet-agent-6.10.1`

I'd also suggest versionlocking the package until a fix is released.

`dnf versionlock puppet-agent`

**Comment by alexjfisher** [comment]
Fixed in recently released puppet 6.12.0

**Comment by blackknight36** [comment]
Appears that this is no longer an issue.  I was able to set up an selinux port on a server running puppet-agent 6.15.0 without any errors.
